src/probe_identity.rs

```diff
@@ -0,0 +1,186 @@
+use color_eyre::eyre::{eyre, Result};
+use color_eyre::Report;
+use log::warn;
+
+use crate::metadata::structs::Probe;
+
+const BMP_PRODUCT_STRING: &str = "Black Magic Probe";
+const BMP_PRODUCT_STRING_LENGTH: usize = BMP_PRODUCT_STRING.len();
+const NATIVE_BPM: &str = "native";
+
+pub enum ProbeIdentity
+{
+    ProbeVersion { name: String, version: String },
+    Probe {name: String},
+    None
+}
+
+fn probe_from_bpm(input: &str) -> Result<String>
+{
+    let opening_paren = input.find('(');
+    let closing_paren = input.find(')');
+
+    match (opening_paren, closing_paren) {
+        (None, None) => Ok(NATIVE_BPM.into()),
+        (Some(opening_paren), Some(closing_paren)) => {
+            if opening_paren > closing_paren {
+                Err(eyre!("'(' is defined after ')'"))
+            } else {
+               Ok(input[opening_paren+1..closing_paren].to_string().to_lowercase())
+            }
+        }
+        (Some(_), None) => Err(eyre!("No ')' defined")),
+        (None, Some(_)) => Err(eyre!("No '(' defined")),
+    }
+}
+
+fn version_from_bpm(input: &str) -> Result<String>
+{
+    let start_index = input.rfind(' ')
+        .ok_or_else(|| eyre!("Invalid version string"))?;
+
+    let version = input[start_index + 1..].to_string();
+
+    if version.trim().is_empty() {
+        return Err(eyre!("Version is empty or only whitespace"));
+    }
+
+    Ok(version)
+}
+
+impl From<String> for ProbeIdentity
+{
+    fn from(identity: String) -> Self
+    {
+        if identity == BMP_PRODUCT_STRING {
+            ProbeIdentity::Probe {
+                name: NATIVE_BPM.into()
+            }
+        } else if identity.starts_with(BMP_PRODUCT_STRING) {
+            let probe_result = probe_from_bpm(&identity[BMP_PRODUCT_STRING_LENGTH..]);
+
+            let probe = match probe_result {
+                Ok(probe) => probe,
+                Err(error) => {
+                    warn!("Error while parsing probe string: {}", error);
+                    return  ProbeIdentity::None
+                }
+            };
+            
+            let version = version_from_bpm(&identity[BMP_PRODUCT_STRING_LENGTH..]);
+
+            match version {
+                Ok(version) => ProbeIdentity::ProbeVersion {
+                    name: probe,
+                    version
+                },
+                Err(error) => {
+                    warn!("Error while parsing version string: {}", error);
+                    ProbeIdentity::None
+                }
+            }            
```

Comment:
This structure is intriguing but seems to throw away work from previous parsing stages, while obfuscating how this is walking the input string.. It looks like the result of `version_from_bpm()` could be `.map()`'d to make it into the shape of the result type for this function for example.